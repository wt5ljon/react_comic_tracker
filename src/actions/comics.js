import uuid from 'uuid';
import database from '../firebase/firebase';

export const addComic = (comic) => {
  return {
    type: 'ADD_COMIC',
    comic
  };
};

export const startAddComic = (comicData = {}) => {
  return (dispatch) => {
    const {
      seriesName = '',
      seriesNumber = undefined,
      storyBy = '',
      artBy = '',
      publicationDate = undefined,
      readStatus = false
    } = comicData;
  const comic = { seriesName, seriesNumber, storyBy, artBy, publicationDate, readStatus };
  database.ref('comics').push(comic).then((ref) => {
    dispatch(addComic({
      id: ref.key,
      ...comic
    }))
  })
  };
};

export const removeComic = (id) => ({
  type: 'REMOVE_COMIC',
  id
});

export const editComic = (id, editedComic) => ({
  type: 'EDIT_COMIC',
  id,
  editedComic
});

export const toggleReadStatus = (id) => ({
  type: 'TOGGLE_READ_STATUS',
  id  
});

export const setComics = (comics) => ({
  type: 'SET_COMICS',
  comics
});

export const startSetComics = () => {
  return (dispatch) => {
    return database.ref('comics')
      .once('value')
      .then((snapshot) => {
        const comics = [];
        snapshot.forEach((childSnapshot) => {
          comics.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });
        dispatch(setComics(comics));
      });
  };
};
