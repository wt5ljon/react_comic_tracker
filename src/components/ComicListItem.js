import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import moment from 'moment';

const ComicListItem = ({
  id, seriesName, seriesNumber, publicationDate, storyBy, artBy
}) => (
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{seriesName}</h3>
      <span className="list-item__sub-title">
        {'Publication Date: '}
        {moment(publicationDate).format('MMMM Do, YYYY')}
      </span>
      <span className="list-item__sub-title">
        {'Story By: '}
        {storyBy}
      </span>
      <span className="list-item__sub-title">
        {'Art By: '}
        {artBy}
      </span>
    </div>
    <h3 className="list-item__data">{seriesNumber}</h3>
  </Link>
);

ComicListItem.propTypes = {
  id: PropTypes.string.isRequired,
  seriesName: PropTypes.string.isRequired,
  seriesNumber: PropTypes.number.isRequired,
  storyBy: PropTypes.string.isRequired,
  artBy: PropTypes.string.isRequired,
  publicationDate: PropTypes.number.isRequired
};

export default ComicListItem;
