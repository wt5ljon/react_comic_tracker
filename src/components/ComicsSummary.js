import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import selectComics from '../selectors/selectComics';

const ComicsSummary = ({ comicCount }) => {
  const comicWord = comicCount === 1 ? 'comicbook' : 'comicbooks';

  return (
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">
          {'Viewing '}
          <span>{comicCount}</span>
          {` ${comicWord}`}
        </h1>
        <div className="page-header__actions">
          <Link className="button" to="/create">Add Comic</Link>
        </div>
      </div>
    </div>
  );
};

ComicsSummary.propTypes = {
  comicCount: PropTypes.number.isRequired
};

const mapStateToProps = (state) => {
  const visibleComics = selectComics(state.comics, state.filters);

  return {
    comicCount: visibleComics.length
  };
};

export default connect(mapStateToProps)(ComicsSummary);
