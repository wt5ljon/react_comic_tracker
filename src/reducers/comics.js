const initialState = [];

const comicsReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'ADD_COMIC':
      return [
        ...state,
        action.comic
      ];
    case 'REMOVE_COMIC':
      return state.filter((comic) => comic.id !== action.id);
    case 'EDIT_COMIC':
      return state.map((comic) => {
        if (comic.id === action.id) {
          return {
            ...comic,
            ...action.editedComic
          };
        } else {
          return comic;
        }
      });
    case 'SET_COMICS':
      return action.comics;
    default:
      return state;
  };
};

export default comicsReducer;