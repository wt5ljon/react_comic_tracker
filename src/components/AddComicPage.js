import React from 'react';
import PropTypes from 'prop-types';
import ReactRouterPropTypes from 'react-router-prop-types';
import { connect } from 'react-redux';
import { startAddComic } from '../actions/comics';
import ComicForm from './ComicForm';

const AddComicPage = ({dispatch, history}) => (
  <div>
    <div className="page-header">
      <div className="content-container">
        <h1 className="page-header__title">Add Comicbook</h1>
      </div>    
    </div>
    <div className="content-container">
      <ComicForm
        onSubmit={(comic) => {
          dispatch(startAddComic(comic));
          history.push('/');
        }}
        buttonText="Save Comicbook"
      />    
    </div>
  </div>
);

AddComicPage.propTypes = {
  dispatch: PropTypes.func.isRequired,
  history: ReactRouterPropTypes.history.isRequired
};

export default connect()(AddComicPage);
