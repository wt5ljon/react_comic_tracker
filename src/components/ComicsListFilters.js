import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { setTextFilter } from '../actions/filters';

const ComicsListFilters = ({dispatch, filters}) => (
  <div>
    <input
      type="text"
      value={filters.text}
      onChange={(e) => {
        dispatch(setTextFilter(e.target.value));
      }}
    />
  </div>
);

ComicsListFilters.propTypes = {
  dispatch: PropTypes.func.isRequired,
  filters: PropTypes.shape({
    text: PropTypes.string.isRequired
  }).isRequired
};

const mapStateToProps = state => ({
  filters: state.filters
});

export default connect(mapStateToProps)(ComicsListFilters);
