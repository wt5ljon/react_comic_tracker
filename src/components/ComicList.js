import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import ComicListItem from './ComicListItem';
import SelectComics from '../selectors/getVisibleComics';

const ComicList = props => {
  const {
    comics
  } = props;

  return (
    <div>
      <h1>Comicbook List</h1>
      {comics.length === 0 && <h2>No Comicbooks to List</h2>}
      {comics.map(comic => (<ComicListItem key={comic.id} {...comic} />))}
    </div>)
};

ComicList.propTypes = {
  comics: PropTypes.arrayOf(PropTypes.object).isRequired
}

const mapStateToProps = state => ({
  comics: SelectComics(state.comics, state.filters)
});

export default connect(mapStateToProps)(ComicList);
