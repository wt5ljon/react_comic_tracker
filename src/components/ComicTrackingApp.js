import React from 'react';
import ComicList from './ComicList';
import ComicsListFilters from './ComicsListFilters';

const ComicTrackingApp = () => {
  return (
    <div>
      <ComicsListFilters />
      <ComicList />
    </div>
  );
};

export default ComicTrackingApp;