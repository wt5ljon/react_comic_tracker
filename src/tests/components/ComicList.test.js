import React from 'react';
import { shallow } from 'enzyme';
import { UComicList } from '../../components/ComicList';
import comics from '../fixtures/comics';

test('should render comic list with comicss', () => {
  const wrapper = shallow(<UComicList comics={comics} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render comic list displaying empty message', () => {
  const wrapper = shallow(<UComicList comics={[]} />);
  expect(wrapper).toMatchSnapshot();
});