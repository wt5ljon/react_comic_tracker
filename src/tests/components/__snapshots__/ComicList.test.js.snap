// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render comic list displaying empty message 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Titles
    </div>
    <div
      className="show-for-desktop"
    >
      Series Title
    </div>
    <div
      className="show-for-desktop"
    >
      Series Number
    </div>
  </div>
  <div
    className="list-body"
  >
    <div
      className="list-item list-item--message"
    >
      <span>
        No Comicbooks
      </span>
    </div>
  </div>
</div>
`;

exports[`should render comic list with comicss 1`] = `
<div
  className="content-container"
>
  <div
    className="list-header"
  >
    <div
      className="show-for-mobile"
    >
      Titles
    </div>
    <div
      className="show-for-desktop"
    >
      Series Title
    </div>
    <div
      className="show-for-desktop"
    >
      Series Number
    </div>
  </div>
  <div
    className="list-body"
  >
    <ComicListItem
      artBy="Ditko"
      id="1"
      key="1"
      publicationDate={0}
      seriesName="Amazing Spider-Man"
      seriesNumber={2}
      storyBy="Lee"
    />
    <ComicListItem
      artBy="Ditko"
      id="2"
      key="2"
      publicationDate={-345600000}
      seriesName="Fantastic Four"
      seriesNumber={13}
      storyBy="Lee"
    />
    <ComicListItem
      artBy="Ditko"
      id="3"
      key="3"
      publicationDate={345600000}
      seriesName="The Avengers"
      seriesNumber={99}
      storyBy="Lee"
    />
  </div>
</div>
`;
